Entry points for review reminders

ericli3690 opened the pull request:
## Purpose / Description
- Created the [EditNotifications] activity, added it to [AndroidManifest]
- Added new string resources for [EditNotifications] to [12-dont-translate.xml]
- Added buttons to launch the [EditNotifications] activity from [DeckPicker], [StudyOptionsFragment], and [NotificationsSettingsFragment]
- Implemented a developer flag in the Developer Options settings menu together with the appropriate gates to ensure the new notifications system only executes when explicitly enabled
- Updated a pre-existing [DeckPickerTest] test to also test the [EditNotifications] button

![image](https://github.com/user-attachments/assets/5f3d5a64-8123-4348-af52-9a995816ad8f)
![image](https://github.com/user-attachments/assets/1e3593a4-ff64-4ccb-9bad-fd712127fa65)

## Fixes
- Part of GSoC 2025: Review Reminders

## Approach
- Currently starting by creating the space within which the new system will function.
- Created a new [EditNotifications] activity and defined the ways it can be opened, as per the following two links in my GSoC proposal: [[1]](https://docs.google.com/document/d/1KMjnm6ol1M_VMAsg9V5Qb4XRZnEON-39gkzLkFe8q2Y/edit?tab=t.0#heading=h.o472euuadswv) [[2]](https://docs.google.com/document/d/1KMjnm6ol1M_VMAsg9V5Qb4XRZnEON-39gkzLkFe8q2Y/edit?tab=t.0#heading=h.4qb4c9dwxqbr)
- The next steps will be to fill in the RecyclerView within [EditNotifications] and to begin trying to create/read/update/delete notification data.

## How Has This Been Tested?
- DeckPickerTest executes successfully.
- Opening EditNotifications through all created routes (long pressing a deck in the DeckPicker screen, clicking the toolbar bell icon in [StudyOptionsFragment], and clicking the app-wide / deck-specific button options in Settings) works and displays expected information.
- When the developer option is disabled, no new changes are accessible.
- Tested on a physical Samsung S23, API 34.

## Learning
- This took me a while longer than I expected because I suddenly ran into some problems first with building and then with executing tests within [DeckPickerTest]. I think Android Studio didn't like running the parameterized tests for some reason and I had to manually edit the build configuration of each individual test before hitting run.
- I initially was trying to figure out how to import a new bell icon and even asked around the Discord server for the process for importing svgs before realizing there was already a bell icon in the resources folder. Whoops.

## Checklist
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [x] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [x] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

ericli3690 commented:
Apologies for the repeated force pushes, my code is failing a few tests. Running the tests on my own machine is really annoying because a lot of them are flaky (in particular I'm running up against this known bug here: #14796). Does anyone know if there's a fix for that issue? Discussion in that thread on fixing it seems to have stalled out.

criticalAY commented on AnkiDroid/src/main/res/xml/preferences_dev_options.xml:
```diff
@@ -109,5 +109,9 @@
             android:key="@string/pref_browser_find_replace"
             android:defaultValue="false"/>
+        <SwitchPreferenceCompat
+            android:title="Experimental notifications"
+            android:key="@string/pref_experimental_notifications"
+            android:defaultValue="false"/>
```
Split it to a separate commit where you are adding it, I would prefer `New notifications` rather than experimental 
and don't hardcode the strings.

ericli3690 replied:
Done!

criticalAY commented on AnkiDroid/src/main/java/com/ichi2/anki/EditNotifications.kt:
```diff
@@ -0,0 +1,95 @@
+import com.ichi2.anki.CollectionManager.withCol
+import com.ichi2.compat.CompatHelper.Companion.getSerializableCompat
+
+class EditNotifications : AnkiActivity() {
```
Do we really need a activity as I think we can do with Fragment and then again split it to different commit

ericli3690 replied:
The problem is that I want this screen to be accessed from and reused in multiple locations, in particular from both the Settings menu and individual decks. Additionally, I also plan to control some of the buttons in the top app bar. I think using an Activity will be simpler and more straightforward, as otherwise I would have to figure out what Activity should host the EditNotificationsFragment, access the top app bar via that host Activity, etc. Is there a particular way you were thinking I could use a Fragment for this screen?

criticalAY commented on AnkiDroid/src/main/res/values/12-dont-translate.xml:
```diff
@@ -26,5 +26,14 @@
+
+    <!-- Experimental Notifications -->
+    <!-- In progress for GSoC 2025 -->
+    <string name="edit_notifications" maxLength="28" comment="Label for the notifications icon menu button that appears in the top bar of the individual deck detailed options menu, and the name of the button option for editing notifications for a deck when a deck is long clicked in the main screen, and the title of the screen that appears when users decide to edit the notifications of a single sp ...
```
Replied on discord this can be marked as resolved

Arthur-Milchior commented:
For the future, in your PR and commit message, please uses backticks instead of square brackets. This way, github shows this as code.

Square bracket is great in kotlin documentation, because that's the way javadoc works and javadoc links to other part of the code. But it's not something that github or most git tools know how to interprets. So instead we should use markdown and not javadoc syntax

I'm not asking you to change the current commits or PR. I don't care enough about this issue to ask you to do extra work here

SanjaySargam commented on AnkiDroid/src/main/res/values/12-dont-translate.xml:
```diff
@@ -26,5 +26,16 @@
+    <!-- In progress for GSoC 2025 -->
+    <!-- TODO: Remove developer flag when the new notification system is stable -->
+    <string name="enable_new_notifications" maxLength="41" comment="Label in the developer options screen that enables the new notifications system.">Enable new notifications</string>
+    <string name="edit_notifications" maxLength="28" comment="Label for the notifications icon menu button that appears in the top bar of the individual deck detailed options menu, and the name of the button option for editing notifications for a deck when a deck is long clicked in the main screen, and the title of the screen that appears when users decide to edit the notifications of a single sp ...
```
This string is not used in this commit. It would be better to either move them to the commit where they are actually used or include their usage in this commit for consistency.

SanjaySargam commented on AnkiDroid/src/main/res/values/12-dont-translate.xml:
```diff
@@ -26,5 +26,16 @@
+    <string name="edit_all_app_wide_notifications" maxLength="41" comment="Label for the notifications settings button which users can click to view and edit notifications that apply to all decks.">Edit all app-wide notifications</string>
+    <string name="app_wide_notifications" comment="Title of the screen that appears when users are viewing and editing notifications that apply to all decks.">App-wide notifications</string>
+    <string name="edit_all_deck_specific_notifications" maxLength="41" comment="Label for the notifications settings button which users can click to view and edit notifications that apply for individual decks.">Edit all deck-specific notifications</string>
+    <string name="deck_specific_notifications" comment="Title of the screen that appears when all deck-specific notifications are being viewed and edited.">Deck-specific notifications</string>
```
(Same here too)
This string is not used in this commit. It would be better to either move them to the commit where they are actually used or include their usage in this commit for consistency.

SanjaySargam commented on AnkiDroid/src/main/res/values/12-dont-translate.xml:
```diff
@@ -26,5 +26,16 @@
+    <string name="enable_new_notifications" maxLength="41" comment="Label in the developer options screen that enables the new notifications system.">Enable new notifications</string>
+    <string name="edit_notifications" maxLength="28" comment="Label for the notifications icon menu button that appears in the top bar of the individual deck detailed options menu, and the name of the button option for editing notifications for a deck when a deck is long clicked in the main screen, and the title of the screen that appears when users decide to edit the notifications of a single sp ...
+    <string name="edit_all_app_wide_notifications" maxLength="41" comment="Label for the notifications settings button which users can click to view and edit notifications that apply to all decks.">Edit all app-wide notifications</string>
+    <string name="app_wide_notifications" comment="Title of the screen that appears when users are viewing and editing notifications that apply to all decks.">App-wide notifications</string>
```
Same here too

ericli3690 commented:
Apologies for the square brackets. I've corrected them to backticks where possible and will continue to do so in the future.

As per discussions on Discord, I've removed basically all edits to 12-dont-translate.xml and am now hardcoding all strings as literals within the code itself. I'll pull them out into a string resource file in August once the feature is stable. The exception is "Edit notifications", which I need to keep as a string resource so that `DeckPickerContextMenu` can accept it as an argument in an enum declaration (it doesn't work if I just pass in the string literal), and so that it can be used as a title in study_options_fragment.xml (the linter gets mad if I don't). For now, I've kept this single resource in 12-dont-translate.xml. Let me know if I should move it somewhere else.

ericli3690 commented:
Ah, something I've just thought of. Should I rename `EditNotifications` to `EditReviewReminders`? I've noticed that I sometimes substitute the word "notifications" for "review reminders" in my code which is probably not good. It might be best to standardize the name of this feature. Is "Review Reminders" the final official English name for this feature? It was the original name of this idea on the GSoC ideas list.

ericli3690 commented on AnkiDroid/src/main/res/menu/study_options_fragment.xml:
```diff
@@ -1,5 +1,11 @@
+    <item
+        android:id="@+id/action_edit_notifications"
+        android:title="@string/edit_notifications_do_not_translate"
+        android:visibility = "gone"
```
For some reason, the rest of the `visibility="gone"` modifiers in this file also all have that weird whitespace. I initially kept it for consistency's sake, but it's now fixed.

ericli3690 commented on AnkiDroid/src/main/res/layout/edit_notifications.xml:
```diff
@@ -0,0 +1,37 @@
+            android:background="?android:attr/colorBackground"
+            android:overScrollFooter="@color/transparent"
+            android:clipToPadding="false"
+            android:paddingBottom="72dp"
```
I initially copied this XML file from card_browser.xml and kept it around since the XML file is mostly still a placeholder at this point. Fixed.

ericli3690 commented on AnkiDroid/src/main/java/com/ichi2/anki/StudyOptionsFragment.kt:
```diff
@@ -351,6 +358,13 @@ class StudyOptionsFragment :
             if (currentContentView == CONTENT_CONGRATS) {
                 menu.findItem(R.id.action_custom_study).isVisible = false
             }
+            // Developer flag: Only show notifications button if new notifications are enabled
```
Ah, good point. I've now removed all the comments of this type.

ericli3690 commented on AnkiDroid/src/main/java/com/ichi2/anki/EditNotifications.kt:
```diff
@@ -0,0 +1,97 @@
+    /**
+     * Mode: The possible modes the EditNotifications activity can be triggered in
+     */
+    private enum class Mode {
```
Hm, I've switched it to `SchedulerScope` (in keeping with the renaming of `EditNotifications` to `ScheduleReminders`). If anyone has a better idea, please let me know.

ericli3690 commented on AnkiDroid/src/main/java/com/ichi2/anki/EditNotifications.kt:
```diff
@@ -0,0 +1,97 @@
+            return intent
+        }
+
+        fun getIntentForSingleDeckSpecificEditing(
```
Oh, I like this version a lot more, good catch! Implemented.

ericli3690 commented on AnkiDroid/src/main/java/com/ichi2/anki/EditNotifications.kt:
```diff
@@ -0,0 +1,97 @@
+ *  this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package com.ichi2.anki
```
Cool, I've created a `reviewreminders` package.

ericli3690 commented on AnkiDroid/src/main/java/com/ichi2/anki/EditNotifications.kt:
```diff
@@ -0,0 +1,97 @@
+import com.ichi2.anki.CollectionManager.withCol
+import com.ichi2.compat.CompatHelper.Companion.getSerializableCompat
+
+class EditNotifications : AnkiActivity() {
```
Good point. How about `ScheduleReminders`?

ericli3690 commented:
As per @lukstbit's recommendation, I've rebased and squashed all my commits together. I've done a major refactoring to switch `EditNotifications` into `ScheduleReminders` and am now using a fragment within `SingleFragmentActivity` rather than creating a whole new activity. The UI has changed marginally (the top bar is now white instead of blue) in keeping with how most other `SingleFragmentActivity` fragments are styled within AnkiDroid. Updated images are in the PR description.

lukstbit commented on AnkiDroid/src/main/java/com/ichi2/anki/StudyOptionsFragment.kt:
```diff
@@ -351,6 +364,11 @@ class StudyOptionsFragment :
             if (currentContentView == CONTENT_CONGRATS) {
                 menu.findItem(R.id.action_custom_study).isVisible = false
             }
+            if (requireContext().sharedPrefs().getBoolean(getString(R.string.pref_new_notifications), false)) {
+                menu.findItem(R.id.action_schedule_reminders).isVisible = true
+            } else {
+                menu.findItem(R.id.action_schedule_reminders).isVisible = false
+            }
```
```suggestion
            val enableNewNotifications = requireContext().sharedPrefs().getBoolean(getString(R.string.pref_new_notifications), false)
            menu.findItem(R.id.action_schedule_reminders).isVisible = enableNewNotifications
```

lukstbit commented on AnkiDroid/src/main/java/com/ichi2/anki/reviewreminders/ScheduleReminders.kt:
```diff
@@ -0,0 +1,115 @@
+                    withCol {
+                        decks.name(did)
+                    }
+                toolbar.subtitle = deckName
```
Combine the two lines if you don't plan to used _deckName_ for other things here:
```

toolbar.subtitle = 
    withCol {...
```

lukstbit commented on AnkiDroid/src/main/res/layout/fragment_schedule_reminders.xml:
```diff
@@ -0,0 +1,43 @@
+            android:id="@+id/toolbar"
+            android:layout_width="match_parent"
+            android:layout_height="wrap_content"
+            app:layout_constraintTop_toTopOf="parent"
```
This is useless when the parent is a LinearLayout.

ericli3690 replied:
Oops. Fixed!

criticalAY commented on AnkiDroid/src/main/java/com/ichi2/anki/reviewreminders/ScheduleReminders.kt:
```diff
@@ -0,0 +1,115 @@
+    private lateinit var mSchedulerScope: SchedulerScope
+
+    /**
+     * SchedulerScope: The possible modes the ScheduleReminders fragment can be triggered in
```
This can be better it doesn't have to start with the class name

criticalAY commented on AnkiDroid/src/main/java/com/ichi2/anki/preferences/NotificationsSettingsFragment.kt:
```diff
@@ -43,6 +46,17 @@ class NotificationsSettingsFragment : SettingsFragment() {
             preferenceScreen.removePreference(requirePreference<SwitchPreferenceCompat>(R.string.pref_notifications_blink_key))
         }
+
+        // If the new review reminders system is enabled, we will not use the legacy system
+        if (requireContext().sharedPrefs().getBoolean(getString(R.string.pref_new_notifications), false)) {
+            displayNewNotificationsSubscreen()
+            return
+        }
+
+        // Otherwise, remove the new review reminders buttons and continue with the legacy system
```
can we move the comment at one place only

criticalAY commented on AnkiDroid/src/main/java/com/ichi2/anki/reviewreminders/ScheduleReminders.kt:
```diff
@@ -0,0 +1,115 @@
+        super.onViewCreated(view, savedInstanceState)
+
+        // Retrieve arguments, update toolbar accordingly
+        mSchedulerScope = requireArguments().getSerializableCompat<SchedulerScope>(EXTRAS_SCOPE_KEY) as SchedulerScope
```
use `schedulerScope` instead of `m`

criticalAY commented on AnkiDroid/src/main/res/menu/study_options_fragment.xml:
```diff
@@ -1,15 +1,21 @@
+        android:id="@+id/action_schedule_reminders"
+        android:title="@string/schedule_reminders_do_not_translate"
+        android:visibility="gone"
+        android:icon="@drawable/ic_notifications"
+        ankidroid:showAsAction="always"/>
     <item
         android:id="@+id/action_rebuild"
         android:title="@string/rebuild_cram_label"
-        android:visibility = "gone"
+        android:visibility="gone"
```
Why this change?

ericli3690 replied:
@lukstbit told me to format it like this for the one XML element I added, and I decided to try making it consistent across the file. But sure, I'll revert these changes.

criticalAY commented on AnkiDroid/src/main/res/menu/study_options_fragment.xml:
```diff
@@ -1,15 +1,21 @@
         android:title="@string/rebuild_cram_label"
-        android:visibility = "gone"
+        android:visibility="gone"
         android:icon="@drawable/ic_refresh_white"
         ankidroid:showAsAction="ifRoom"/>
     <item
         android:id="@+id/action_empty"
         android:title="@string/empty_cram_label"
-        android:visibility = "gone"
+        android:visibility="gone"
```
Why this change? revert

criticalAY commented on AnkiDroid/src/main/res/menu/study_options_fragment.xml:
```diff
@@ -26,5 +32,5 @@
         android:id="@+id/action_unbury"
         android:title="@string/unbury"
-        android:visibility = "gone"/>
+        android:visibility="gone"/>
```
Revert

lukstbit reviewed (approved):
LGTM as a starting point.

ericli3690 commented:
Refactored. UI images in the PR description have been updated. Looking forward to reviews!

criticalAY commented on AnkiDroid/src/main/java/com/ichi2/anki/deckpicker/DeckPickerViewModel.kt:
```diff
@@ -177,6 +178,11 @@ class DeckPickerViewModel :
         }
+
+    fun scheduleReviewReminders(deckId: DeckId) =
+        launchCatchingIO {
+            flowOfDestination.emit(ScheduleRemindersDestination(false, deckId))
+        }
```
Do we need `launchCatchingIO` is any IO operation being performed?

criticalAY replied:
Then why not

```
 viewModelScope.launch { ... }
```

ericli3690 replied:
Ah, good point. Fixed!

criticalAY commented on AnkiDroid/src/main/java/com/ichi2/anki/preferences/DevOptionsFragment.kt:
```diff
@@ -159,6 +159,15 @@ class DevOptionsFragment : SettingsFragment() {
         }
+
+        /**
+         * The new review reminders system changes things in the settings screen,
+         * so we need to reload the screen immediately.
+         */
```
Can do better with doc `things` isn't clear

criticalAY reviewed (changes requested):
Great, just minor requests

SanjaySargam commented on AnkiDroid/src/main/java/com/ichi2/anki/reviewreminders/ScheduleReminders.kt:
```diff
@@ -0,0 +1,104 @@
+
+        val toolbar =
+            view.findViewById<MaterialToolbar>(R.id.toolbar).apply {
+                setTitle("Schedule reminders")
```
You may use `R.string.schedule_reminders_do_not_translate`

ericli3690 commented:
Thanks Ashish and Sanjay! Updated.

ericli3690 commented:
- Replaced `launchCatchingIO` with `viewModelScope.launch`

criticalAY reviewed (approved):
Cool!!!

SanjaySargam reviewed (commented):
LGTM 🚀

ericli3690 commented on AnkiDroid/src/main/java/com/ichi2/anki/preferences/HeaderFragment.kt:
```diff
@@ -120,7 +137,16 @@ class HeaderFragment :
+                    searchConfiguration
+                        .indexItem()
+                        .withKey(activity.getString(R.string.pref_review_reminders_screen_key))
+                        .withTitle("Review reminders")
```
I was told by Arthur to hardcode my strings since this feature is still in development. I plan on going back through all my commits and pulling out all the strings in late August!

SanjaySargam replied:
ok cool

ericli3690 commented:
Cool! Thanks for the review. Since there are two approvals, I can mark this as Pending Merge, right?

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/StudyOptionsFragment.kt:
```diff
@@ -351,6 +364,9 @@ class StudyOptionsFragment :
                 menu.findItem(R.id.action_custom_study).isVisible = false
             }
+            // Use new review reminders system if enabled
+            val enableNewReviewReminders = requireContext().sharedPrefs().getBoolean(getString(R.string.pref_new_notifications), false)
```
"enable" seems to indicate that you are doing an action to enable it.
I would have called it "newReviewReminderIsEnabled"

ericli3690 replied:
Good point, I'll change this in my next PR.

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/reviewreminders/ScheduleReminders.kt:
```diff
@@ -0,0 +1,104 @@
+     * This private variable is only used for the deck-specific editing scope.
+     * In global scope, it is set to -1.
+     */
+    private var did: DeckId = -1L
```
I'd have used the constant `NOT_FOUND_DECK_ID` just so that it's the same everywhere, but not blocking on it. Same below

ericli3690 replied:
Yep, I was planning to do that. I'll wait for https://github.com/ankidroid/Anki-Android/pull/18364 to be merged first, though, since it makes more sense to declare the constant there. Thanks for the review!

ericli3690 commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DeckPickerContextMenu.kt:
```diff
@@ -75,6 +76,9 @@ class DeckPickerContextMenu : AnalyticsDialogFragment() {
             if (!dyn) {
                 add(DeckPickerContextMenuOption.EDIT_DESCRIPTION)
             }
+            if (requireContext().sharedPrefs().getBoolean(getString(R.string.pref_new_notifications), false)) {
```
Gosh, every time I think I've found all the cool utility functions y'all have made for this project, I learn of yet another. This is much cleaner, thanks!

ericli3690 commented on AnkiDroid/src/main/java/com/ichi2/anki/reviewreminders/ScheduleReminders.kt:
```diff
@@ -0,0 +1,104 @@
+            }
+
+        // Retrieve arguments, update toolbar accordingly
+        isInGlobalScope = requireArguments().getBoolean(EXTRAS_GLOBAL_SCOPE_KEY, true)
```
Ah, this is interesting. I'll look into refactoring this, might open a short PR dedicated to fixing up these little suggestions, or I might lump these changes in with the RecyclerView PR once I get that wrapped up. Thanks!

ericli3690 commented:
I've incorporated the above suggestions into my code and will open up a PR for the updates once https://github.com/ankidroid/Anki-Android/pull/18364 is merged.
